### ISSUE TITLE
Wrong instance handles

### ISSUE THREAD

**Issue by trittsv** [eclipse-cyclonedds/cyclonedds-cxx#379]
Hi, we are using CycloneDDS-CXX 0.10.2.

The instance handles are not matching with the receiving instance handles received by listeners.

Publisher.cpp
```cpp
class ExampleWriterListener : public virtual dds::pub::DataWriterListener<HelloWorldData::Msg> {
public:
    virtual void on_publication_matched (. . .){
          std::cout << "Incoming handle:"<< status.last_subscription_handle()<< std::endl; 
    }
}
int main() {
    // . . .
    dds::pub::DataWriter<HelloWorldData::Msg> writer(. . .);

    std::cout << "publisher writer instance handle:" << writer.instance_handle() << std::endl;
}
```
and the other side (subscriber.cpp):
```cpp
class ExampleReaderListener : public virtual dds::sub::DataReaderListener<HelloWorldData::Msg> {
public:      
    virtual void on_subscription_matched (. . .){
            std::cout << "Incoming handle:"<< status.last_publication_handle()<< std::endl;
      }
}
int main() {
        // . . .
    dds::sub::DataReader<HelloWorldData::Msg> reader(. . .);

    std::cout << "Subscriber reader instance handle:" << reader.instance_handle() << std::endl;
}
```

Publisher output
```bash
publisher writer instance handle: 143130644491087548
# on_publication_matched
Incoming handle: 13344768980678218717 # i would expect "3255376580424972098"
```
Subscriber output
```bash
Subscriber reader instance handle: 3255376580424972098
# on_subscription_matched
Incoming handle: 10063450733852295754 # i would expect "143130644491087548"
```

We are wondering why they the numbers are not identical (for example using RTI the numbers are matching).

@eboasson Should the numbers match in your opinion?

**Comment by eboasson** [comment]
This is correct behaviour for Cyclone. In the DDS spec, the "publication_handle" you get in the sample info is only guaranteed to be valid in the context of the `get_matched_publication_data` operation (see, e.g., 2.2.2.5.3.33 get_matched_publication_data and 2.2.2.5.5 SampleInfo class).

In Cyclone I've tried to give a few more guarantees, because there are actually two ways to get this data: one is by calling `get_matched_publication_data`, the other is by looking at the contents of a reader for the "DCPSPublication" topic. Each instance in the latter data reader is associated with an instance handle as well — or at least it returns one in the sample info and there is no statement that it is (allowed to be) meaningless in this case. According to the spec, the only way of creating a "DCPSPublication" is by going through `get_builtin_subscriber` and `lookup_datareader`, and so, according to the spec, these readers for built-in topics do not have to be "real" data readers.

Clearly, all of these special cases lead to a fairly messy set of behaviours and guarantees, and besides this last point implies that while you can always create many readers of your own topics, an implementation myopically following this part of the spec only allows you a single reader for a built-in topic — and in consequence, only a single set of sample and instance states.

So in Cyclone, I chose to do this differently. All readers (including built-in readers) are perfectly ordinary readers, the built-in topics somehow magically exist, all `instance_handle`s and `publication_handle`s line up inside one process (well, inside one instantiation of the spec, if you trick it into [creating two instances in one process](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/manual/options.md#cycloneddsdomaindiscoveryexternaldomainid) those will be completely independent) and *all* instance handles are handled through the same mechanism.

And for various independent reasons those instance handles deliberately look like random garbage — they're actually generated by a 64-bit block cipher in counter mode with each instance of Cyclone using a random key.

This mapping of GUID to instance handle is stable at least as long as the GUID is referenced by a known entity (local or remote) or by a reader for a built-in topic. It may be longer, both because deleting them is not completely instantaneous, and because of another design feature (but then you're slowly moving into the implementation-defined behaviours).

This other design feature, which today unfortunately only applies to the C code, is that the instance handles of any pair of instances with the same key value in *different* readers are guaranteed to be the same. That should help doing equi-joins on the output of data readers, and if you decide to use a hash-table, the best hash function will be the identity function. But unfortunately, because of reasons, the C++ type support code expands in such a way that this detail doesn't apply there — it can be changed, and hopefully it will be, but it isn't a super high priority.

On a final note, I do wonder how RTI would guarantee these handles are the same if there are multiple writer processes: then surely there is a possibility of these two writer processes happen to create a data writer that locally have the same instance handle (*), and then the subscriber side can no longer use the same one for both. Obviously there are a variety of tricks you can use to minimise the likelihood of running into this collision, and then guaranteeing that the two will be the same in the absence of a collision (and make sure everything works out correctly if there is a collision). But it wouldn't really help an application program ...

Granted, it would be nice to the poor human who is looking at those identifiers!

(*) Unless their `publication_handle` is the GUID (or a ≥128-bit blob) but then all their "regular" instance handles also need to be that large. Maybe they are, it is definitely allowed and it would be in line with the DDSI "key hash" thing, but in that case, one should worry about MD5 collisions. Cyclone never uses DDSI key hashes for anything because of that risk of collisions. We even have a [test](https://github.com/eclipse-cyclonedds/cyclonedds/blob/e52fda8ff136fd76df73a902ba24cfc97343fd8b/src/core/ddsc/tests/instance_handle.c#L136) to sanity check we handle them correctly.
